File: python/composio/utils/sensitive_file_upload_paths.py
"""Block accidental upload of local files from well-known secret/credential locations."""

from __future__ import annotations

import re
import typing as t
from pathlib import Path

# Path components that indicate a sensitive directory anywhere in a resolved local path.
BUILTIN_FILE_UPLOAD_PATH_DENY_SEGMENTS: t.Tuple[str, ...] = (
    ".ssh",
    ".aws",
    ".azure",
    ".gnupg",
    ".kube",
    ".docker",
    ".claude",  # may contain API keys and project context read by assistants
    ".password-store",
    "keychains",
)

_SECRET_LIKE_BASENAME = re.compile(r"^(\.env(\.|$)|\.netrc$|\.pgpass$)", re.IGNORECASE)
_DEFAULT_PRIVATE_KEY_BASENAME = re.compile(
    r"^id_(rsa|ed25519|ecdsa|dsa|ecdsa_sk)(\.old)?$", re.IGNORECASE
)


def _normalize_path_segments(file_path: t.Union[str, Path]) -> t.List[str]:
    p = Path(file_path).expanduser()
    try:
        resolved = p.resolve()
    except OSError:
        resolved = p
    parts = [part for part in resolved.as_posix().split("/") if part]
    if parts and parts[0].endswith(":"):
        # Windows path: "C:/Users/..." -> drop drive letter segment
        parts = parts[1:]
    return parts


def _get_block_reason(
    file_path: t.Union[str, Path],
    additional_deny_segments: t.Optional[t.Sequence[str]] = None,
) -> t.Optional[str]:
    extra = [s.strip() for s in (additional_deny_segments or []) if s and s.strip()]
    deny: t.Set[str] = {s.lower() for s in BUILTIN_FILE_UPLOAD_PATH_DENY_SEGMENTS}
    deny.update(s.lower() for s in extra)

    segments = _normalize_path_segments(file_path)
    # Case-insensitive segment match: lower each path component once per check.
    for seg, key in zip(segments, (s.lower() for s in segments), strict=True):
        if key in deny:
            return f'path segment "{seg}" is in the sensitive file upload denylist'

    if not segments:
        return None
    basename = segments[-1]
    if _SECRET_LIKE_BASENAME.search(basename) or _DEFAULT_PRIVATE_KEY_BASENAME.match(
        basename
    ):
        return (
            f'file name "{basename}" looks like a credential, env, or private key file'
        )
    if basename.lower() == "credentials":
        return 'file name "credentials" is often used for cloud/API credential stores'
    return None


def is_blocked_sensitive_file_upload_path(
    file_path: t.Union[str, Path],
    additional_deny_segments: t.Optional[t.Sequence[str]] = None,
) -> bool:
    return _get_block_reason(file_path, additional_deny_segments) is not None


def assert_safe_local_file_upload_path(
    file_path: t.Union[str, Path],
    *,
    enabled: bool = True,
    additional_deny_segments: t.Optional[t.Sequence[str]] = None,
) -> None:
    """Raise SensitiveFilePathBlockedError if *file_path* matches the denylist."""
    if not enabled:
        return
    reason = _get_block_reason(file_path, additional_deny_segments)
    if reason:
        from composio.exceptions import SensitiveFilePathBlockedError

        raise SensitiveFilePathBlockedError(
            f"Refusing to upload: {reason}. "
            "Set sensitive_file_upload_protection=False on Composio if you must "
            "(not recommended), or use a copy outside sensitive locations."
        )

File: ts/packages/core/src/utils/sensitiveFileUploadPaths.ts
/**
 * Blocks accidental upload of local files from well-known secret/credential locations
 * during auto file upload (and {@link getFileDataAfterUploadingToS3}).
 *
 * This is the single canonical guard shared by the core SDK and downstream
 * packages (e.g. `@composio/cli`). Filesystem access goes through the `#platform`
 * abstraction so this module stays free of static `node:*` imports and is safe to
 * re-export from the package root (edge/workerd builds included).
 */
import { platform } from '#platform';
import { ComposioSensitiveFilePathBlockedError } from '../errors/FileModifierErrors';

/**
 * Path segments (a single path component) that indicate a sensitive directory when
 * they appear anywhere in a resolved local path. Matching follows the case sensitivity
 * of the filesystem containing that path.
 */
export const BUILTIN_FILE_UPLOAD_PATH_DENY_SEGMENTS: readonly string[] = [
  '.ssh',
  '.aws',
  '.azure',
  '.gnupg',
  '.kube',
  '.docker',
  '.claude', // may contain API keys and project context read by assistants
  '.password-store',
  'keychains', // e.g. ~/Library/Keychains
];

const SECRET_LIKE_BASENAME = /^(\.env(\.|$)|\.netrc$|\.pgpass$)/i;
/** Default SSH private key basenames (public keys like id_rsa.pub are allowed). */
const DEFAULT_PRIVATE_KEY_BASENAME = /^id_(rsa|ed25519|ecdsa|dsa|ecdsa_sk)(\.old)?$/i;

const splitSegments = (aPath: string): string[] => aPath.split(/[/\\]+/).filter(Boolean);

/**
 * Returns normalized path segments, resolving symlinks when the path exists.
 *
 * Both the written path and the symlink-resolved one are returned, because a
 * denied segment can be hidden by a symlink in either direction:
 *
 *   - `~/innocent-name` -> `~/nested/.aws/creds` hides `.aws` from the written
 *     path, so the denylist has to see the resolved one.
 *   - `~/.claude` -> `/state/claude` hides `.claude` from the *resolved* path,
 *     so the denylist also has to see the written one. Dotfile managers
 *     (chezmoi, stow, yadm) and containerised home directories produce exactly
 *     this layout.
 *
 * Matching only the resolved path silently turns the denylist off for the
 * second case.
 */
function normalizePath(filePath: string): {
  resolvedPath: string;
  segments: string[];
  writtenSegments: string[];
} {
  const absolute = platform.resolvePath(filePath);
  let resolved = absolute;
  try {
    if (platform.existsSync(absolute)) {
      resolved = platform.realpathSync(absolute);
    }
  } catch {
    // If realpath fails (e.g. race), use resolved path
  }
  return {
    resolvedPath: resolved,
    segments: splitSegments(resolved),
    writtenSegments: splitSegments(absolute),
  };
}

/**
 * True if the path is under a built-in or extra deny segment, or the basename
 * looks like an env / netrc / default SSH private key name.
 */
export function isBlockedSensitiveFileUploadPath(
  filePath: string,
  additionalDenySegments?: string[]
): boolean {
  return getSensitiveFileUploadPathBlockReason(filePath, additionalDenySegments) != null;
}

function getSensitiveFileUploadPathBlockReason(
  filePath: string,
  additionalDenySegments?: string[]
): string | null {
  const { resolvedPath, segments, writtenSegments } = normalizePath(filePath);
  const isCaseSensitive = platform.isFileSystemCaseSensitive?.(resolvedPath) ?? true;
  const normalizeSegment = (segment: string) => (isCaseSensitive ? segment : segment.toLowerCase());
  const deny = new Set(
    [
      ...BUILTIN_FILE_UPLOAD_PATH_DENY_SEGMENTS,
      ...(additionalDenySegments ?? []).map(s => s.trim()).filter(Boolean),
    ].map(normalizeSegment)
  );

  for (const candidate of [segments, writtenSegments]) {
    const segmentsForMatch = isCaseSensitive ? candidate : candidate.map(normalizeSegment);
    for (let i = 0; i < candidate.length; i++) {
      if (deny.has(segmentsForMatch[i]!)) {
        return `path segment "${candidate[i]}" is in the sensitive file upload denylist`;
      }
    }
  }

  // Same two directions as the segment scan: `~/.env -> /state/config` hides a
  // denied basename from the resolved path, and a symlink pointing *at* a
  // credential file hides it from the written one.
  const basenames = [segments.at(-1), writtenSegments.at(-1)].filter(
    (name): name is string => !!name
  );
  for (const basename of basenames) {
    if (SECRET_LIKE_BASENAME.test(basename) || DEFAULT_PRIVATE_KEY_BASENAME.test(basename)) {
      return `file name "${basename}" looks like a credential, env, or private key file`;
    }
    if (basename.toLowerCase() === 'credentials') {
      return 'file name "credentials" is often used for cloud/API credential stores';
    }
  }
  return null;
}

/**
 * Default remediation hint appended to the block message. Assumes an SDK caller
 * that exposes `sensitiveFileUploadProtection`. Callers without such an opt-out
 * (e.g. `@composio/cli`) should pass their own `remediation` so the message does
 * not advertise an option the caller cannot honor.
 */
const DEFAULT_REMEDIATION =
  `To upload from this path anyway, set sensitiveFileUploadProtection: false on Composio ` +
  `(not recommended) or use a copy outside sensitive locations.`;

/**
 * @throws {ComposioSensitiveFilePathBlockedError} if the path is not allowed
 */
export function assertSafeFileUploadPath(
  filePath: string,
  options?: { additionalDenySegments?: string[]; remediation?: string }
): void {
  const reason = getSensitiveFileUploadPathBlockReason(filePath, options?.additionalDenySegments);
  if (reason) {
    const remediation = options?.remediation ?? DEFAULT_REMEDIATION;
    throw new ComposioSensitiveFilePathBlockedError(
      `Refusing to upload: ${reason}. ${remediation}`,
      {
        meta: { filePath, reason },
      }
    );
  }
}
